include/ofi_util.h

```diff
@@ -1328,15 +1328,29 @@ static inline void ofi_cq_err_memcpy(uint32_t api_version,
 				     const struct fi_cq_err_entry *prov_buf)
 {
 	size_t size;
+	char *err_buf_save;
+	size_t err_data_size;
 
-	if (FI_VERSION_GE(api_version, FI_VERSION(1, 20)))
-		size = sizeof(struct fi_cq_err_entry);
-	else if (FI_VERSION_GE(api_version, FI_VERSION(1, 5)))
-		size = sizeof(struct fi_cq_err_entry_1_1);
-	else
-		size = sizeof(struct fi_cq_err_entry_1_0);
+	if (FI_VERSION_LT(api_version, FI_VERSION(1, 5))) {
+		memcpy(user_buf, prov_buf, sizeof(struct fi_cq_err_entry_1_0));
+	} else {
+		err_buf_save = user_buf->err_data;
+		err_data_size = MIN(user_buf->err_data_size,
+				    prov_buf->err_data_size);
+
+		if (FI_VERSION_GE(api_version, FI_VERSION(1, 20)))
+			size = sizeof(struct fi_cq_err_entry);
+		else
+			size = sizeof(struct fi_cq_err_entry_1_1);
+
+		memcpy(user_buf, prov_buf, size);
 
-	memcpy(user_buf, prov_buf, size);
+		if (err_data_size)
+			memcpy(err_buf_save, prov_buf->err_data, err_data_size);
+
+		user_buf->err_data = err_buf_save;
```

Comment:
If input `err_data_size` is 0, the output `err_data` always refers to the provider returned value, regardless of the API version.

Reply:
Looking at the fi_cq manpage, it only states that err_data is set to a provider returned value if err_data_size == 0 or API version < 1.5. For err_data_size > 1 and API >= 1.5, the manpage says that on output, the provider will fill in the user err_data with any provider specific data.

Reply:
That's right: if err_data_size == 0 `or` API version < 1.5
However, here you set `user_buf->err_data` back to the input value even if err_data_size is 0. This line needs to be guarded with `if (err_data_size > 0)`.

Reply:
More precisely, should save `user_buf->err_data_size` and use it here to guard the assignment, and report `prov_buf->err_data_size` on the next line if `user_buf->err_data_size` is 0.

Reply:
Ok. I think I have this fixed now.